### PR TITLE
Pull request for #86: Integrate make_provenance into SConstruct

### PR DESCRIPTION
**Pull request by arosenbe**
@stanfordquan, can you review my work from #86. 
* Changes the naming conventions for our paths variables (to easily incorporate user_config.yaml)
* Store pointers to external inputs in yamls under the `external` key (this is just a hard coded naming convention, couldn't think of an easy way to abstract) 
* Implements the new `record_dir` function from gslab-econ/gslab_python#131 on all the pointers under `external`

Tandem to gslab-econ/gslab_python#132

### CODE REVIEW

File: analysis/SConstruct [hunk 1]
```diff
@@ -25,17 +24,25 @@ env = Environment(ENV = {'PATH': os.environ['PATH']},
...
 # Extensions to be used when scanning for source files in BuildLyx.
 env.EXTENSIONS = ['.eps', '.pdf', '.lyx']
 SourceFileScanner.add_scanner('.lyx', Scanner(gs.misc.lyx_scan, recursive = True))
-# Load paths
-env['PATHS'] = yaml.load(open('config_global.yaml', 'rU'))
 # Load Stata executable if Stata builder is defined
 if 'BuildStata' in env['BUILDERS'].keys():
-   env['stata_executable'] = misc.load_yaml_value('config_user.yaml', 'stata_executable')
+    env['stata_executable'] = gs.misc.load_yaml_value('config_user.yaml', 'stata_executable')
+# Load global and user config file, record contents of their external directories
```
Comment (stanfordquan) [inline analysis/SConstruct]: This is a lot of code to put in `SConstruct`. Can we abstract this out and then pass in arguments with `['global', 'user']` and `external` as defaults?

File: analysis/source/descriptive/descriptive.py [hunk 2]
```diff
@@ -3,14 +3,14 @@
...
+    data  = np.genfromtxt('%s/data.txt' % paths_global['build']['prepare_data'], skip_header = 1)
 
-    with open('%s/table.txt' % PATHS['build']['descriptive'], 'w') as f:
+    with open('%s/table.txt' % paths_global['build']['descriptive'], 'w') as f:
         f.write('<tab:table>\n')
         f.write('%s\n%.3f\n%d\n%d' % (np.mean(data), np.std(data, ddof = 1), np.max(data), np.min(data)))
 
     plt.hist(data)
-    plt.savefig('%s/plot.eps' % PATHS['build']['descriptive'])
+    plt.savefig('%s/plot.eps' % paths_global['build']['descriptive'])
```
Comment (stanfordquan) [inline analysis/source/descriptive/descriptive.py]: I see the issue with carrying around two `paths` variables. My instinct is to write a wrapper function `read_config()` that read both of these into a `path` variable or whatever it's called and then have keys for `global` and `user`.

File: analysis/release/state_of_external.log [hunk 3]
```diff
@@ -0,0 +1,384 @@
...
+paper_slides/source/paper/online_appendix.lyx|3610
+paper_slides/source/paper/paper.lyx|5236
+paper_slides/source/paper/SConscript|677
+paper_slides/source/paper/text.lyx|2522
+paper_slides/source/paper/text.tex|1300
+paper_slides/source/tables/SConscript|233
+paper_slides/source/tables/table.lyx|4082
+paper_slides/source/talk/SConscript|237
+paper_slides/source/talk/slides.lyx|4697
+config/scons/scons-local-2.5.1/SCons/__init__.py|1634
```
Comment (stanfordquan) [inline analysis/release/state_of_external.log]: we definitely want to exclude this directory

Comment (stanfordquan) [reply analysis/release/state_of_external.log]: or maybe not I'm not that definite about it on second thoughts

File: analysis/config_global.yaml [hunk 4]
```diff
@@ -8,5 +8,9 @@ source:
   prepare_data: source/prepare_data
   descriptive: source/descriptive
 
+# External Direcotries to record
```
Comment (stanfordquan) [inline analysis/config_global.yaml]: directories

File: analysis/config_global.yaml [hunk 5]
```diff
@@ -8,5 +8,9 @@ source:
...
+external:
+  self: ..
```
Comment (stanfordquan) [inline analysis/config_global.yaml]: I'm not sure this is how we want to demonstrate this feature. I was imagining adding this to `paper_slides` and then have it work on `input`.

File: analysis/SConstruct (hunk 1 above)
Comment (arosenbe) [reply analysis/SConstruct]: I agree that it's a lot of code. I'm not sure where to abstract it to. I originally had it as part of record_dir, but that's no longer appropriate. I've dumped it as two new blocks in configuration.py. That seems like a reasonable resting place for it. Let me know though.

File: analysis/release/state_of_external.log (hunk 3 above)
Comment (arosenbe) [reply analysis/release/state_of_external.log]: This record is just a result of me not knowing which external directory I should point to. I froze and went with `..`, the entire template. I think there's no actual value in recording the state of template before a run, other than as demonstration. But you're right at https://github.com/gslab-econ/template/pull/87#discussion_r137432001 that we can come up with a better demo.

File: analysis/release/state_of_external.log [hunk 6]
```diff
@@ -0,0 +1,384 @@
+===================================
+
+ self
+
+===================================
+
```
[the remark this answers is not in the document]
Comment (arosenbe) [reply analysis/release/state_of_external.log]: This log uses the same formatting as state_of_repo.log. The headings are literally produced by the same function. Do you want me to change the style for both logs?

[the remark this answers is not in the document]
Comment (stanfordquan) [reply analysis/release/state_of_external.log]: In this case we're going to have a lot of these headers, where as in state_of_repo you only have one so there's no need to search. I think it's good to change here, whether we should change in state_of_repo is up to you

File: analysis/source/descriptive/descriptive.py (hunk 2 above)
Comment (arosenbe) [reply analysis/source/descriptive/descriptive.py]: This work but it means that the PATHS variables we load in the scripts have a different structure than the PATHS environment variable we use in SCons.

File: analysis/release/state_of_external.log (hunk 6 above)
[the remark this answers is not in the document]
Comment (arosenbe) [reply analysis/release/state_of_external.log]: There's two things in state of repo, and I'm going to change both.

**Comment by arosenbe** [comment]
@stanfordquan, hold off reviewing this until the next push for gslab-econ/gslab_python#132.

**Comment by arosenbe** [comment]
@stanfordquan, push has gone through, can you look at my changes?

File: analysis/SConstruct [hunk 7]
```diff
@@ -18,12 +18,13 @@ gs.log.start_log(mode, vers)
 env = Environment(ENV = {'PATH': os.environ['PATH']}, 
                   IMPLICIT_COMMAND_DEPENDENCIES = 0,
                   BUILDERS = {
-                              # 'BuildMatlab': Builder(action = gs.build_matlab),
+                              'BuildMatlab': Builder(action = gs.build_matlab),
```
Comment (stanfordquan) [inline analysis/SConstruct]: why uncomment?

Comment (arosenbe) [reply analysis/SConstruct]: Was testing. Will recomment.

**Comment by arosenbe** [comment]
Will write change log in summary commit.

File: analysis/SConstruct (hunk 7 above)
Comment (arosenbe) [reply analysis/SConstruct]: Whoops, forgot to do this before merging. Done here though https://github.com/gslab-econ/template/commit/e5659df655beab7ca7db3fb39336873814414d49.

**Pull request merged**